### PR TITLE
[ADD] module_auto_update: Create module

### PR DESCRIPTION
**Pull request by hughesbm**
Automate module updates by comparing module directory checksums.

Completes RFC #790

### CODE REVIEW

File: module_auto_update/__manifest__.py
Comment (lasley) [inline module_auto_update/__manifest__.py]: `LasLabs` if on the clock. You go in Contributors section of ReadMe using LasLabs email.

File: module_auto_update/__manifest__.py
Comment (lasley) [inline module_auto_update/__manifest__.py]: relicense to LGPL-3, this was OK'd in the RFC https://github.com/OCA/server-tools/issues/790#issuecomment-299258642

File: module_auto_update/__manifest__.py
Comment (lasley) [inline module_auto_update/__manifest__.py]: LGPL, apply to all

File: module_auto_update/README.rst
Comment (lasley) [inline module_auto_update/README.rst]: LGPL

File: module_auto_update/README.rst
Comment (lasley) [inline module_auto_update/README.rst]: LGPL

File: module_auto_update/README.rst
Comment (lasley) [inline module_auto_update/README.rst]: lgpl

File: module_auto_update/__init__.py
Comment (lasley) [inline module_auto_update/__init__.py]: Newline at end of file

File: module_auto_update/__manifest__.py
Comment (lasley) [inline module_auto_update/__manifest__.py]: Remove empty keys

File: module_auto_update/data/cron_data.xml
Comment (lasley) [inline module_auto_update/data/cron_data.xml]: `<odoo noupdate="1">` & can remove data tags

File: module_auto_update/data/cron_data.xml
Comment (lasley) [inline module_auto_update/data/cron_data.xml]: Simple strings should just be in the XML text node. Eval is for python code:

`<function>check_upgrade_availability</function>`

File: module_auto_update/data/cron_data.xml
Comment (lasley) [inline module_auto_update/data/cron_data.xml]: `<field name="model">ir.module.module</field>`

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: Remove the spacing - it's a bit much IMO

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: Trailing comma

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: s/upgradable/upgradeable

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: Filters ftw:

``` python
upgradable_modules = upgradable_modules.filtered(
    lambda r: r.checksum_dir != r.checksum_installed
)
upgradable_modules.write({
    'upgrade_available': True,
})
```

It's more lines, but it is less McCabe score & performs the write in one database query. Odoo ActiveRecord pattern may seem nice, but it hits the db each time - vs a write can update a crapton of records with the same data at once.

ActiveRecord in SQL:
``` sql
UPDATE my_table SET col = val WHERE id = 1;
UPDATE my_table SET col = val WHERE id = 2;
UPDATE my_table SET col = val WHERE id = 3;
UPDATE my_table SET col = val WHERE id = 4;
```

Write:

``` sql
UPDATE my_table SET col = val WHERE id in [1, 2, 3, 4];
```

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]:
```
r.write({
    'checksum_installed': r.checksum_dir,
    'upgrade_available': False,
})
```

File: module_auto_update/views/module_schedule_upgrade.xml
Comment (lasley) [inline module_auto_update/views/module_schedule_upgrade.xml]: can remove data tag

File: module_auto_update/views/module_schedule_upgrade.xml
Comment (lasley) [inline module_auto_update/views/module_schedule_upgrade.xml]: newline at EOF

File: module_auto_update/views/module_views.xml
Comment (lasley) [inline module_auto_update/views/module_views.xml]: no data

File: module_auto_update/wizard/module_upgrade.py
Comment (lasley) [inline module_auto_update/wizard/module_upgrade.py]: kill the spaces

**Review by lasley — CHANGES_REQUESTED** [review]
Excellent first PR @hughesbm!

Comments inline

File: module_auto_update/models/module.py
Comment (yajo) [reply module_auto_update/models/module.py]: Can't this be into `button_immediate_upgrade` directly?

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: You don't need this line.

File: module_auto_update/security/ir.model.access.csv
Comment (yajo) [inline module_auto_update/security/ir.model.access.csv]: Do you really need this file?

**Review by yajo — APPROVED** [review]
Code OK (apart from @lasley's comments)

File: module_auto_update/models/module.py
Comment (hughesbm) [reply module_auto_update/models/module.py]: @Yajo It's also called from `upgrade_module` in `base.module.upgrade`; `button_immediate_upgrade` isn't used for applying scheduled upgrades.

File: module_auto_update/security/ir.model.access.csv
Comment (hughesbm) [reply module_auto_update/security/ir.model.access.csv]: I guess not! Thanks! 👍

**Comment by hughesbm** [comment]
The requested changes have been addressed. I still need to add tests, however.

File: module_auto_update/__init__.py
Comment (tafaRU) [inline module_auto_update/__init__.py]: This line is not needed as `tests` is automatically inspected by Odoo.

**Comment by lasley** [comment]
Ping @jjscarafia

File: module_auto_update/__manifest__.py
Comment (yajo) [inline module_auto_update/__manifest__.py]: I'd use `Extra Tools`

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: Could we just set state=to upgrade?

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: If we use state to track upgradeable addons, then you should remove `to upgrade` from here.

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: I'm not sure... does `api.depends` work if the method is not marked as a `compute` method on some field?

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: Same about depends here

File: module_auto_update/views/module_views.xml
Comment (yajo) [inline module_auto_update/views/module_views.xml]: If we use the state as said above, could we move all the checksum computing into `update_list()`, so we don't have to create additional actions, menu items and methods? This way, if the admin uses the standard system to upgrade the addons list, he will always update their checksums too. It seems like what he'd expect to happen.

File: module_auto_update/data/cron_data.xml
Comment (yajo) [inline module_auto_update/data/cron_data.xml]: I see no sense on having 2 crons. Checking the upgrades without upgrading makes no sense, as upgrading without checking what's to upgrade neither. Besides, you could have a race condition where there are more than 1 cron workers. Please have a Python method that does all and run it in a single cron.

File: module_auto_update/wizard/module_upgrade.py
Comment (yajo) [inline module_auto_update/wizard/module_upgrade.py]: Add `to install` too?

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: One question... If we would have addons directory full of `.pyc` or `.pyo` files, and under a containerized frozen environment where those files are autoregenerated on each build, would that produce a different checksum? I guess it should be the same because if the pycompiled file is based on the same input python file, it should produce the same output, but did you check this? Or is it possible to ignore such files?

**Review by yajo — CHANGES_REQUESTED** [review]
I feel this needs a little more simplification.

File: module_auto_update/models/module.py
Comment (lasley) [reply module_auto_update/models/module.py]: I think you're correct here Jairo. The `pyc` files contain a `moddate` header, which is the date of compilation as I understand it. Due to this, the hash can and will change when the file is recomputed, regardless of whether there were actually any changes.

Luckily `checksumdir` has an `excluded_extension` key, which interestingly enough has `pyc` in it as the example:

```
sha256hash = dirhash(directory, 'sha256', excluded_extensions=['pyc'])
```

IMO we should use an array declared at class level to determine the extensions that should be excluded. We could also use a method, but I think that's a bit overengineered

Comment (yajo) [reply module_auto_update/models/module.py]: The simplest balance between simplicity and configurability is an unexisting `ir.config_parameter` with sane defaults:

```python
exclude = self.env["ir.config_parameter"].get_param(
    "module_auto_update.checksum_excluded_extensions", 
    "pyc,pyo").split(",")
```

This code with a little note in the configuration section of README should do the trick.

Comment (lasley) [reply module_auto_update/models/module.py]: @hughesbm - here's an example of [ir.config_param](https://github.com/OCA/website/blob/10.0/website_form_recaptcha/data/ir_config_parameter_data.xml)

Comment (lasley) [reply module_auto_update/models/module.py]: Ummm but don't copy whatever the hell I did with the tab/space mixture there. Spaces4Lyfe!

File: module_auto_update/data/cron_data.xml
Comment (lasley) [reply module_auto_update/data/cron_data.xml]: I'm good with automating the whole thing. Let's hide the `Apply Scheduled Upgrades` button in the UI then as well.

File: module_auto_update/models/module.py
Comment (lasley) [reply module_auto_update/models/module.py]: It does not; you can only edit computed fields with depends methods. They technically run, but the changes aren't stored to DB because you are in the cache layer.

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: `@api.depends('name')`

File: module_auto_update/models/module.py
Comment (yajo) [reply module_auto_update/models/module.py]: Yeah, although you don't really need to provide a `ir.config_param` record if you add a fallback value as in the code I put above.

File: module_auto_update/models/module.py
Comment (yajo) [reply module_auto_update/models/module.py]: Then this decorator makes no sense

File: module_auto_update/data/cron_data.xml
Comment (yajo) [reply module_auto_update/data/cron_data.xml]: wouldn't it be better to extend that method and make the cron call it instead?

File: module_auto_update/models/module.py
Comment (lasley) [reply module_auto_update/models/module.py]: Agreed - I was mainly just explaining for my newb 😉

File: module_auto_update/data/cron_data.xml
Comment (lasley) [reply module_auto_update/data/cron_data.xml]: Ohhhh yes, I like that better. That way we can trigger a manual update without the cron

**Comment by hughesbm** [comment]
In addition to the requested changes, I made a fairly large change to how `checksum_installed` is stored. The strategy of overriding module upgrade methods to store the checksum completely ignored module installation; modules would be immediately scheduled for upgrade after installation.

To deal with this, this module now uses the `@api.depends` decorator to watch for module state changes, and recomputes `checksum_installed` when state changes to `'installed'`. Cancellation of upgrades and uninstalls also changes the state to 'installed', however, so those methods were overridden to prevent undesired behavior.

File: module_auto_update/README.rst
Comment (lasley) [inline module_auto_update/README.rst]: with the desired extensions listed as comma separated values.

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: This field must not be computed. It only has to be updated when addon gets updated too, but it is a normal char field.

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: You don't need these kind of weird hacks if `checksum_installed` is not a computed field.

**Review by yajo — CHANGES_REQUESTED** [review]
The way you are using the computation of `checksum_installed` seems to go against the purpose of such kind of fields.

If you want to check when an addon's state changes to `installed`, but skip doing things under some situations, you should override `write` and `create` methods, and put something like this inside them:

```python
def write(self, vals):
  if vals.get("state") == "installed" and not self.env.context.get("retain_checksum_installed"):
    for one in self:
      one.checksum_installed = one.checksum_dir
  return super(Class, self).write(vals)
```

Also, on the methods where you need to retain the installed checksum:

```python
def medthod_X(self):
  return super(Class, self.with_context(retain_checksum_installed=True)).method_X()
```

This would follow the normal Odoo framework logic and hooks to make all work as expected and be written as expected 😉

File: module_auto_update/models/module.py
Comment (hughesbm) [reply module_auto_update/models/module.py]: Yeah, this is a lot better. 😆  Thanks!

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: First, create the record, and then call to `result._store_checksum_installed(vals)`, or it will have no effect.

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: First write the record and then call the other method, or it could have undesired results.

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: I'd also check `"checksum_installed" not in vals`, to avoid double writes if the field is being manually written.

File: module_auto_update/wizard/module_upgrade.py
Comment (yajo) [reply module_auto_update/wizard/module_upgrade.py]: Just noting this seems OK now.

File: module_auto_update/models/module.py
Comment (yajo) [reply module_auto_update/models/module.py]: Just noting this seems OK now.

**Review by yajo — CHANGES_REQUESTED** [review]

**Comment by yajo** [comment]
BTW, you have conflicts.

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py]: Set it to `False`, which is the standard in Odoo when we have an empty field.

File: module_auto_update/models/module.py
Comment (yajo) [inline module_auto_update/models/module.py:23]: I feel this one should be read only, mostly now that you are exposing it in the UI.
You should also add `string` and `help` attributes to both fields for the same reason, to let the user know what's this.

**Review by yajo — CHANGES_REQUESTED** [review]

File: module_auto_update/models/module.py
Comment (hughesbm) [reply module_auto_update/models/module.py:23]: Makes sense. Does exposing these fields in the UI have value, or should I take it out again? I keep wavering on this.

Comment (lasley) [reply module_auto_update/models/module.py:23]: The value would be for the human. If a human cares, expose it - otherwise don't. In this case, I'd say don't.

File: module_auto_update/hooks.py
Comment (lasley) [inline module_auto_update/hooks.py:9]: This isn't tested. Also I don't think it's implemented correctly, because coverage should have a hit regardless of test status for hooks.

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: trailing comma

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: trailing comma

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: Simplify this conjoined if by moving `not self.env.context` above, inverse the logic, and return.

File: module_auto_update/models/module.py
Comment (lasley) [reply module_auto_update/models/module.py]: `self.write({'checksum_installed': False})` and kill the loop

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: trailing comma

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: trailing comma

File: module_auto_update/models/module.py
Comment (lasley) [inline module_auto_update/models/module.py]: no trailing comma. The point of the trailing comma is to allow for an option to be added without editing multiple lines in the diff. In the instance of a single line, this is a null point

File: module_auto_update/tests/__init__.py
Comment (lasley) [inline module_auto_update/tests/__init__.py]: Newline at EOF

File: module_auto_update/wizard/module_upgrade.py
Comment (lasley) [inline module_auto_update/wizard/module_upgrade.py]: trailing comma

**Review by lasley — CHANGES_REQUESTED** [review]

**Comment by hughesbm** [comment]
Runbot is currently failing when creating the odoo instance, not due to changes here as far as I can tell.

**Comment by lasley** [comment]
Runbot fail due to https://github.com/OCA/maintainer-quality-tools/pull/464

**Review by lasley — APPROVED** [review]

**Review by elicoidal — APPROVED** [review]

**Review by moylop260 — APPROVED** [review]

**Comment by elicoidal** [comment]
@yajo please review again (I think your remark was attended)

**Review by yajo — APPROVED** [review]

**Comment by yajo** [comment]
It looks ready to merge now, just wondering if we could get rid of 505af1e77f312760b51b277a276862325760f2f0 and put everything in a single commit, seems to add no value at my eyes (maybe @jjscarafia has an opinion that we must respect) 😉

**Comment by lasley** [comment]
> just wondering if we could get rid of 505af1e and put everything in a single commit

Agreed - this is pretty much a rewrite at this point. Only a bit of boilerplate was carried over from what I can tell.

**Comment by hughesbm** [comment]
Everything is now in a single commit. Thanks for the fantastic code reviews! 😄

File: module_auto_update/__manifest__.py
Comment (yajo) [inline module_auto_update/__manifest__.py]: I was just about to merge, but saw this. Please do not set it to autoinstall. Also, add dependency on `base` module.

**Review by yajo — CHANGES_REQUESTED** [review]

File: module_auto_update/__manifest__.py
Comment (hughesbm) [reply module_auto_update/__manifest__.py]: Thanks! I have made the requested changes and squashed.

Comment (hughesbm) [reply module_auto_update/__manifest__.py]: Apparently those changes weren't as consequence-free as I expected and a test is now failing. Working on it now.

**Comment by hughesbm** [comment]
@Yajo The changes you requested are done, and I fixed the test that broke as a result. I'll clean up the commits after approval. Sorry about the mess, I got too hasty trying to wrap this up quickly.

**Review by yajo — APPROVED** [review]
Great! That's what I love about good tests like yours, they fail whey they must! 😉

**Comment by lasley** [comment]
Congrats on your first module Brent! Not very often your first one gets to be this fun or popular!

**Comment by Yenthe666** [comment]
Hooray! I'll test this next week too. Kudos!

**Pull request merged**
